rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,849 @@
+- This should be supported by most HSMs through the `CKM_RSA_PKCS_OAEP` and
+  `CKM_SHA256` mechanisms which are usable when calling the `C_DecryptInit`
+  function exposed by PKCS#11.
+
+Each time there is a change to the active keys, the set of public keys will
+also be assigned to `session_recording_config.status.encryption_keys` to be
+used by nodes throughout the cluster.
+
+In order to avoid unintended automatic deletion, keys provisioned for encrypted
+session recording will be tagged with a different label where applicable. This
+will prevent older auth servers from deleting this new keytype and allow new
+auth servers to identify which keys are no longer in use before deletion.
+
+Auth servers will need to create a watcher for `Put` events against the
+`recording_encryption` resource in order to ensure continued access to the
+correct decryption keys. This will most often come into play during key
+rotations and keystore migrations when new keys enter the active set and old
+keys shift into the rotated set.
+
+#### Keystore Managed
+
+When key management is left to the keystore, Teleport makes no attempt at
+generating encryption keys. Instead it will use the labels defined in
+`session_recording_config.encryption.active_keys` and
+`session_recording_config.encryption.rotated_keys` to query keys from the
+configured keystore. The key IDs are cached by a fingerprint of their public
+keys and and the active public keys are written to the
+`session_recording_config.status.encryption_keys` list. A fingerprint in the
+context of this RFD is a base64 encoded `SHA256` hash of the public key in
+PKIX ASN.1 DER form. The auth server will refresh its keys whenever it's
+notified of `session_recording_config` changes and also periodically to capture
+any new keys created with the same labels.
+
+It's important to note that historical recordings will only be accessible if
+all auth servers have access to the correct keys. It is the responsibility of
+the keystore and Teleport admins to ensure that keys are not lost and that new
+auth servers entering the cluster have access to the appropriate key material
+to decrypt historical recordings.
+
+It is expected that all recording encryption key pairs are `RSA4096` regardless
+of whether they are Teleport managed or keystore managed.
+
+### Key Sharing
+
+For keys not managed by Teleport, key sharing is left to the administrator to
+facilitate.
+
+In order for all auth servers in a cluster to replay all recordings, they will
+need access to the same keys. There are common cases where keys are easily
+shared, such as a software key or an AWS KMS key that all auth servers have
+access to. This RFD proposes that recording encryption should assume shared
+access to the same keys and keystores from all auth servers in a cluster. For
+networked HSMs, this will require replacing the `host_uuid` label applied to
+keys with a shared label signalling recording encryption.
```

Comment:
If all auth servers have access to the REKs why do we need to keep track of unaccessible keys and individual key rotations and multiple active keys at once?

Reply:
Keeping track is just a failsafe in case that assumption breaks down because of a bad auth server configuration. Multiple active keys should only be possible during key rotations before the completion or rollback RPC is called, otherwise there should just be a single active key.